### PR TITLE
Track subworkflow invocations after main invocation is scheduled

### PR DESCRIPTION
**Pull request by mvdbeek**
Otherwise we'd attempt to download results too early. That's how I discovered
https://github.com/galaxyproject/galaxy/pull/16705.

### CODE REVIEW

File: planemo/galaxy/activity.py [hunk 1]
```diff
@@ -208,33 +208,25 @@ def _execute(  # noqa C901
...
+        final_invocation_state, job_state, error_message = wait_for_invocation_and_jobs(
+            ctx,
+            invocation_id=invocation_id,
+            history_id=history_id,
+            user_gi=user_gi,
+            no_wait=no_wait,
+            polling_backoff=polling_backoff,
+        )
+        if final_invocation_state not in ("ok", "skipped"):
+            msg = f"Failed to run workflow, at least one job is in [{final_invocation_state}] state."
```
Comment (nsoranzo) [inline planemo/galaxy/activity.py:222]:
```suggestion
            msg = f"Failed to run workflow [{workflow_id}], at least one job is in [{final_invocation_state}] state."
```

File: planemo/galaxy/activity.py [hunk 2]
```diff
@@ -722,9 +714,53 @@ def get_dict_from_workflow(gi, workflow_id):
     return gi.workflows.export_workflow_dict(workflow_id)
 
 
-def _wait_for_invocation(ctx, gi, history_id, workflow_id, invocation_id, polling_backoff=0):
+def wait_for_invocation_and_jobs(
```
Comment (nsoranzo) [inline planemo/galaxy/activity.py:717]: Should we also have this implemented in BioBlend?
Once it's in a BioBlend release, we could simplify the code here (and maybe in some Galaxy tests as well).

File: planemo/galaxy/activity.py [hunk 3]
```diff
@@ -722,9 +714,53 @@ def get_dict_from_workflow(gi, workflow_id):
...
+    # TODO: hook in invocation["messages"]
+    error_message = ""
+    job_state = "ok"
+    try:
+        final_invocation_state = _wait_for_invocation(ctx, user_gi, invocation_id, polling_backoff)
+        assert final_invocation_state == "scheduled"
+    except Exception as e:
+        ctx.vlog(f"Problem waiting on invocation: {str(e)}")
+        summarize_history(ctx, user_gi, history_id)
+        error_message = "Final invocation state is [%s]" % final_invocation_state
```
Comment (nsoranzo) [inline planemo/galaxy/activity.py:732]:
```suggestion
        error_message = f"Final state of invocation {invocation_id} is [{final_invocation_state}]"
```

File: planemo/galaxy/activity.py [hunk 4]
```diff
@@ -722,9 +714,53 @@ def get_dict_from_workflow(gi, workflow_id):
...
+
+    ctx.vlog("Final invocation state is [%s]" % final_invocation_state)
```
Comment (nsoranzo) [inline planemo/galaxy/activity.py:734]:
```suggestion
    ctx.vlog(f"Final state of invocation {invocation_id} is [{final_invocation_state}]")
```

File: planemo/galaxy/activity.py [hunk 5]
```diff
@@ -722,9 +714,53 @@ def get_dict_from_workflow(gi, workflow_id):
...
+
+    if not no_wait:
+        job_state = _wait_for_invocation_jobs(ctx, user_gi, invocation_id, polling_backoff)
+        if job_state not in ("ok", "skipped"):
+            msg = "Failed to run workflow, at least one job is in [%s] state." % job_state
```
Comment (nsoranzo) [inline planemo/galaxy/activity.py:739]:
```suggestion
            msg = f"Failed to run workflow, at least one job is in [{job_state}] state."
```

File: planemo/galaxy/activity.py [hunk 6]
```diff
@@ -722,9 +714,53 @@ def get_dict_from_workflow(gi, workflow_id):
...
+                        invocation_id=step["subworkflow_invocation_id"],
+                        history_id=history_id,
+                        user_gi=user_gi,
+                        no_wait=no_wait,
+                        polling_backoff=polling_backoff,
+                    )
+                    if final_invocation_state != "scheduled" or job_state not in ("ok", "skipped"):
+                        return final_invocation_state, job_state, error_message
+
+        ctx.vlog("All invocation and subworkflow invocations states are 'ok'")
```
Comment (nsoranzo) [inline planemo/galaxy/activity.py:757]:
```suggestion
        ctx.vlog(f"The final state of all jobs and subworkflow invocations for invocation [{invocation_id}] is 'ok'")
```

File: planemo/galaxy/activity.py (hunk 2 above)
Comment (mvdbeek) [reply planemo/galaxy/activity.py:717]: I don't know, there's many ways to do it, and none seem optimal ... I'd probably rather include subworkflow invocation jobs on the Galaxy side ?

**Review by nsoranzo — APPROVED** [review]

**Pull request merged**